README.md
graphitesend
============

[![Build Status](https://travis-ci.org/daniellawrence/graphitesend.png?banch=master)](https://travis-ci.org/daniellawrence/graphitesend?branch=master)
![](https://img.shields.io/pypi/dm/graphitesend.svg)
![](https://img.shields.io/pypi/dw/graphitesend.svg)
![](https://img.shields.io/pypi/dd/graphitesend.svg)
[![Changelog](http://allmychanges.com/p/python/graphitesend/badge/)](http://allmychanges.com/p/python/graphitesend/?utm_source=badge)
[![Coverage Status](https://coveralls.io/repos/daniellawrence/graphitesend/badge.png)](https://coveralls.io/r/daniellawrence/graphitesend)

Easy python bindings to write to Carbon ( Re-write of carbonclient).


Read The Docs
-----------
[graphitesend.rtfd.org](http://graphitesend.readthedocs.org/en/latest/)


Blog posts
-----------
[dansysadm.com](http://dansysadm.com/blog/sending_data_to_graphte_from_python.html)

Example Scripts
----------------
The github repo of [graphitesend-examples](https://github.com/daniellawrence/graphitesend-examples)
has lots of examples using graphitesend to grab data from your local linux system.


Usage Example
--------------

Very basic sending of a metric called metric with a value of 45

````python
>>> import graphitesend
>>> graphitesend.init()
>>> graphitesend.send('metric', 45)
>>> graphitesend.send('metric2', 55)
````

The above would send the following metric to graphite

    system.localhostname.metric 45 epoch-time-stamp
    system.localhostname.metric2 55 epoch-time-stamp


Cleaning up the interface and using a group of cpu to alter the metric prefix

````python
>>> import graphitesend
>>> g = graphitesend.init(group='cpu')
>>> g.send('metric', 45)
>>> g.send('metric2', 55)
````

The above would send the following metric to graphite

    system.localhostname.cpu.metric 45 epoch-time-stamp
    system.localhostname.cpu.metric2 55 epoch-time-stamp


Using a different prefix (other then system.hostname)

````python
>>> import graphitesend
>>> g = graphitesend.init(prefix='apache.rc')
>>> g.send('404', 4)
>>> g.send('200', 500)
````

The above would send the following metric to graphite

    apache.rc.localhostname.404 4 epoch-time-stamp
    apache.rc.localhostname.200 500 epoch-time-stamp
    
To get rid of `localhostname`, set another argument — `system_name`

````python
>>> import graphitesend
>>> g = graphitesend.init(prefix='apache.rc', system_name='')
>>> g.send('404', 4)
>>> g.send('200', 500)
````

The above would send the following metric to graphite

    apache.rc.404 4 epoch-time-stamp
    apache.rc.200 500 epoch-time-stamp


Sending a dict()

````python
    >>> import graphitesend
    >>> g = graphitesend.init()
    >>> g.send_dict({'metric': 45, 'metric2': 55})
````

Sending a list()

````python
    >>> import graphitesend
    >>> g = graphitesend.init()
    >>> g.send_list([('metric', 45), ('metric2', 55)])
````

Sending a list(), with a custom timestamp for all metric-value pairs

````python
    >>> import graphitesend
    >>> g = graphitesend.init()
    >>> g.send_list([('metric', 45), ('metric2', 55)], timestamp=12345)
````

Sending a list(), with a custom timestamp for each metric-value pairs

````python
    >>> import graphitesend
    >>> g = graphitesend.init()
    >>> g.send_list([('metric', 45, 1234), ('metric2', 55, 1234)])
````

Learning? Use dryrun.
----------------------

With dryrun enabled the data will never get sent to a remote service, it will
just print out what would have been sent.

````python
    >>> import graphitesend
    >>> g = graphitesend.init(dryrun=True)
    >>> g.send_list([('metric', 45, 1234), ('metric2', 55, 1234)])
````

Example: the init()
----------------

Set a metric prefix (Default arg)
````python
>>> g = graphitesend.init('prefix')
>>> print g.send('metric', 1)
sent 34 long message: prefix.metric 1.000000 1365068929
````

set a metric prefix using kwargs
````python
>>> g = graphitesend.init(prefix='prefix')
>>> print g.send('metric', 2)
sent 34 long message: prefix.metric 2.000000 1365068929
````

Squash any dots in the hostname
````
>>> g = graphitesend.init(fqdn_squash=True)
>>> print g.send('metric', 3)
sent 56 long message: systems.host_example_com.metric 2.00000 1365069029
````
 
view the default prefix, hardset systems. then followed by the name of the
host that execute the send().
````python
>>> g = graphitesend.init()
>>> print g.send('metric', 3)
sent 44 long message: systems.<system_name>.metric 3.000000 1365069029
````

Set a suffix, handy if you have a bunch of timers or percentages
````python
>>> g = graphitesend.init(suffix='_ms')
>>> print g.send('metric', 4)
sent 47 long message: systems.<system_name>.metric_ms 4.000000 1365069100
````

set a system_name if your submitting results for a different system
````python
>>> g = graphitesend.init(system_name='othersystem')
>>> print g.send('metric', 5)
sent 47 long message: systems.othersystem.metric 5.000000 1365069100
````

Lowercase all the metric names that are send to the graphite server.
````python
>>> g = graphitesend.init(lowercase_metric_names=True)
>>> print g.send('METRIC', 6)
sent 47 long message: systems.<hostname>.metric 6.000000 1365069100
````


Set a group name, handy if you just parsed iostat and want to prefix all the 
metrics with iostat, after its already in the <system_name> directory.
````python
>>> g = graphitesend.init(group='groupname')
>>> print g.send('metric', 6)
sent 54 long message: systems.<system_name>.groupname.metric 6.000000 136506924
````

Connect to a different graphite server
````python
>>> graphitesend.init(graphite_server='graphite.example.com')
````

Connect to a different graphite server port
````python
>>> graphitesend.init(graphite_port=2003)
````


Send async messages
````python
>>> graphitesend.init(asynchronous=True)
````


Change connect timeout (default 2)
````python
>>> graphitesend.init(timeout_in_seconds=5)
````


CLI
------------

Just added -- A cli script that allows for anything to send metrics over to 
graphite (not just python).

The usage is very simple you need to give the command a metric and a value.

````sh
	$ graphitesend name.of.the.metric 666
````

Send more\* then 1 metric and value

````sh
	$ graphitesend name.of.the.metric 666
	$ graphitesend name.of.the.other_metric 2
````

\* Call it 2 times ;)

Installing
----------

*pip*

````sh
$ pip install graphitesend
````

or

*source*

````sh
$ git clone git://github.com/daniellawrence/graphitesend.git
$ cd graphitesend
$ python ./setup.py install
````


Porcelain Overview
==================

init
-----  
Create the module instance of GraphiteSend.

send
-----
Make sure that we have an instance of the GraphiteClient. 
Then send the metrics to the graphite server.

send_dict
---------
Make sure that we have an instance of the GraphiteClient.
Then send the metrics to the graphite server.

reset
-----
Disconnect from the graphite server and destroy the module instance.


TCP vs UDP
==========

There is a new branch for UDP support called 'udp and tcp'.
TCP will continue to be the default with UDP as an option


CHANGELOG
0.10.0
------

* Alexandre Bonnetain: Merge pull request #75 from daniellawrence/feature/integer-metric-name
* Alexandre Bonnetain: Handle integer metric name
* Alexandre Bonnetain: Merge pull request #73 from daniellawrence/formatter-file
* Alexandre Bonnetain: Move the formatter's code to its own file
* Alexandre Bonnetain: Merge pull request #72 from daniellawrence/fix/license-2017
* Alexandre Bonnetain: Set gevent to 1.1.0 for Python 2.6
* Alexandre Bonnetain: Update license file for 2017
* Alexandre Bonnetain: Simpler formatting code (#69)
* Dmitry S. Vlasov: add exponential backoff logic for autoreconnect (#66)
* Daniel Lawrence: minor
* Chris Carr: os.uname not supported on windows, use platform.uname
* Alexandre Bonnetain: Merge branch 'factor_formatter' of https://github.com/zillow/graphitesend into master
* Daniel: lock flake8 - due to python2.6 support (#68)
* Daniel: adding py35 test runner (#63)
* Jonathan Ultis: change whitespace in comments to satisfy flake8
* Jonathan Ultis: remove redundant timestamp set
* Jonathan Ultis: remove the untested simple graphite formatter to avoid annoying coverall
* Jonathan Ultis: fix flakes
* Jonathan Ultis: Factor the formatter out of the GraphiteClient.
* kubauk: Fixing issue for Python3. Dispatch method uses methods that attempt to encode a bytes object when it had already. (#60)
* Nate Tangsurat: fix: Init can use relative imports (#54)
* Daniel: Merge pull request #58 from daniellawrence/fix/remove_monkeypatch
* Daniel Lawrence: getting fix/remove_monkeypatch ready
* Alexandre Bonnetain: Merge pull request #56 from daniellawrence/fix/test_clean_metric_name
* Alexandre Bonnetain: Add tests for the case where clean_metric_name is set to false
* Daniel: Merge pull request #55 from e4r7hbug/fix/tox
* Nate Tangsurat: fix: Change asynchronous dependencies for tox
* Alexandre Bonnetain: Pythonic adjustement
* Alexandre Bonnetain: Monkey patch during test execution because it's not in the library anymore
* Alexandre Bonnetain: Enforce the gevent monkey patching if the user activate asynchronoucity
* Alexandre Bonnetain: Warn the user that he should monkey patch sockets to use asynchronoucity
* Alexandre Bonnetain: Remove monkey patching

0.7.0
-----
	* Added python2.6 support, minor formatting change + unitest2
	* Added send message again when socket is empty (@e4r7hug)

0.6.1
-----
	* Added optional clean_metric_name (@Shir0kamii)

0.6.0
-----
	* Happy new year
	* Added autoreconnect (Thanks  @Shir0kamii)

0.5.0
-----
	* Added asynchronous support via gevent (default: False)
	* Added timeout_in_seconds (default: 2)
	* Better test coverage
	* Minor PEP8 fixes

0.4.0
-----
	* Switched to tox for testing
	* Switched to a socket server for testing
	* Adding coverage test
	* Added coveralls.io link
	* Few pep8/flake fixes
	* Fixed missing pickle client

0.3.5
-----

	* Added CHANGELOG
	* Purge generated debian/* files from git


graphitesend/graphitesend.py
#!/usr/bin/env python

try:
    import gevent
except ImportError:
    gevent = False

import pickle
import socket
import struct
import time
import random

from .formatter import GraphiteStructuredFormatter

_module_instance = None

default_graphite_pickle_port = 2004
default_graphite_plaintext_port = 2003
default_graphite_server = 'graphite'

VERSION = "0.10.0"


class GraphiteSendException(Exception):
    pass


class GraphiteClient(object):
    """
    Graphite Client that will setup a TCP connection to graphite.

    :param prefix: string added to the start of all metrics
    :type prefix: Default: "systems."
    :param graphite_server: hostname or ip address of graphite server
    :type graphite_server: Default: graphite
    :param graphite_port: TCP port we will connect to
    :type graphite_port: Default: 2003
    :param debug: Toggle debug messages
    :type debug: True or False
    :param group: string added to after system_name and before metric name
    :param system_name: FDQN of the system generating the metrics
    :type system_name: Default: current FDQN
    :param suffix: string added to the end of all metrics
    :param lowercase_metric_names: Toggle the .lower() of all metric names
    :param fqdn_squash: Change host.example.com to host_example_com
    :type fqdn_squash: True or False
    :param dryrun: Toggle if it will really send metrics or just return them
    :type dryrun: True or False
    :param timeout_in_seconds: Number of seconds before a connection is timed out.
    :param asynchronous: Send messages asynchronouly via gevent (You have to monkey patch sockets for it to work)
    :param clean_metric_name: Does GraphiteClient needs to clean metric's name
    :type clean_metric_name: True or False
    It will then send any metrics that you give it via
    the .send() or .send_dict().

    You can also take advantage of the prefix, group and system_name options
    that allow you to setup default locations where your whisper files will
    be kept.
    eg.
    ( where 'linuxserver' is the name of the localhost)

    .. code-block:: python

      >>> init().prefix
      systems.linuxserver.

      >>> init(system_name='remote_host').prefix
      systems.remote_host.

      >>> init(group='cpu').prefix
      systems.linuxserver.cpu.

      >>> init(prefix='apache').prefix
      apache.

    """

    def __init__(self, prefix=None, graphite_server=None, graphite_port=2003,
                 timeout_in_seconds=2, debug=False, group=None,
                 system_name=None, suffix=None, lowercase_metric_names=False,
                 connect_on_create=True, fqdn_squash=False,
                 dryrun=False, asynchronous=False, autoreconnect=False,
                 clean_metric_name=True):
        """
        setup the connection to the graphite server and work out the
        prefix.

        This allows for very simple syntax when sending messages to the
        graphite server.

        """

        # If we are not passed a host, then use the graphite server defined
        # in the module.
        if not graphite_server:
            graphite_server = default_graphite_server
        self.addr = (graphite_server, graphite_port)

        # If this is a dry run, then we do not want to configure a connection
        # or try and make the connection once we create the object.
        self.dryrun = dryrun
        if self.dryrun:
            self.addr = None
            graphite_server = None
            connect_on_create = False

        # Only connect to the graphite server and port if we tell you too.
        # This is mostly used for testing.
        self.timeout_in_seconds = int(timeout_in_seconds)
        if connect_on_create:
            self.connect()

        self.debug = debug
        self.lastmessage = None

        self.asynchronous = False
        if asynchronous:
            self.asynchronous = self.enable_asynchronous()
        self._autoreconnect = autoreconnect

        self.formatter = GraphiteStructuredFormatter(prefix=prefix, group=group,
                                                     system_name=system_name, suffix=suffix,
                                                     lowercase_metric_names=lowercase_metric_names, fqdn_squash=fqdn_squash,
                                                     clean_metric_name=clean_metric_name)

    @property
    def prefix(self):
        '''Backward compat - access to the properties on the default formatter
        deprecated - use the formatter directly for this type of muckery.
        '''
        return self.formatter.prefix

    @property
    def suffix(self):
        '''Backward compat - access to properties on the default formatter
        deprecated - use the formatter directly for this type of muckery.
        '''
        return self.formatter.suffix

    @property
    def lowercase_metric_names(self):
        '''Backward compat - access to properties on the default formatter
        deprecated - use the formatter directly for this type of muckery.
        '''
        return self.formatter.lowercase_metric_names

    def connect(self):
        """
        Make a TCP connection to the graphite server on port self.port
        """
        self.socket = socket.socket()
        self.socket.settimeout(self.timeout_in_seconds)
        try:
            self.socket.connect(self.addr)
        except socket.timeout:
            raise GraphiteSendException(
                "Took over %d second(s) to connect to %s" %
                (self.timeout_in_seconds, self.addr))
        except socket.gaierror:
            raise GraphiteSendException(
                "No address associated with hostname %s:%s" % self.addr)
        except Exception as error:
            raise GraphiteSendException(
                "unknown exception while connecting to %s - %s" %
                (self.addr, error)
            )

        return self.socket

    def reconnect(self):
        self.disconnect()
        self.connect()

    def autoreconnect(self, sleep=1, attempt=3, exponential=True, jitter=5):
        """
        Tries to reconnect with some delay:

        exponential=False: up to `attempt` times with `sleep` seconds between
        each try

        exponential=True: up to `attempt` times with exponential growing `sleep`
        and random delay in range 1..`jitter` (exponential backoff)


        :param sleep: time to sleep between two attempts to reconnect
        :type sleep: float or int
        :param attempt: maximal number of attempts
        :type attempt: int
        :param exponential: if set - use exponential backoff logic
        :type exponential: bool
        :param jitter: top value of random delay, sec
        :type jitter: int

        """

        p = 0

        while attempt is None or attempt > 0:
            try:
                self.reconnect()
                return True
            except GraphiteSendException:

                if exponential:
                    p += 1
                    time.sleep(pow(sleep, p) + random.randint(1, jitter))
                else:
                    time.sleep(sleep)

                attempt -= 1

        return False

    def clean_metric_name(self, metric_name):
        """
        Make sure the metric is free of control chars, spaces, tabs, etc.
        """
        return self.formatter.clean_metric_name(metric_name)

    def disconnect(self):
        """
        Close the TCP connection with the graphite server.
        """
        try:
            self.socket.shutdown(1)

        # If its currently a socket, set it to None
        except AttributeError:
            self.socket = None
        except Exception:
            self.socket = None

        # Set the self.socket to None, no matter what.
        finally:
            self.socket = None

    def _dispatch_send(self, message):
        """
        Dispatch the different steps of sending
        """

        if self.dryrun:
            return message

        if not self.socket:
            raise GraphiteSendException(
                "Socket was not created before send"
            )

        sending_function = self._send
        if self._autoreconnect:
            sending_function = self._send_and_reconnect

        try:
            if self.asynchronous and gevent:
                gevent.spawn(sending_function, message)
            else:
                sending_function(message)
        except Exception as e:
            self._handle_send_error(e)

        return "sent {0} long message: {1}".format(len(message), message[:75])

    def _handle_send_error(self, error):
        if isinstance(error, socket.gaierror):
            raise GraphiteSendException(
                "Failed to send data to %s, with error: %s" %
                (self.addr, error))

        elif isinstance(error, socket.error):
            raise GraphiteSendException(
                "Socket closed before able to send data to %s, "
                "with error: %s" %
                (self.addr, error)
            )

        else:
            raise GraphiteSendException(
                "Unknown error while trying to send data down socket to %s, "
                "error: %s" %
                (self.addr, error)
            )

    def _send(self, message):
        """
        Given a message send it to the graphite server.
        """

        self.socket.sendall(message.encode("ascii"))

    def _send_and_reconnect(self, message):
        """Send _message_ to Graphite Server and attempt reconnect on failure.

        If _autoreconnect_ was specified, attempt to reconnect if first send
        fails.

        :raises AttributeError: When the socket has not been set.
        :raises socket.error: When the socket connection is no longer valid.
        """
        try:
            self.socket.sendall(message.encode("ascii"))
        except (AttributeError, socket.error):
            if not self.autoreconnect():
                raise
            else:
                self.socket.sendall(message.encode("ascii"))

    def _presend(self, message):
        """
        Complete any message alteration tasks before sending to the graphite
        server.
        """
        return message

    def send(self, metric, value, timestamp=None, formatter=None):
        """
        Format a single metric/value pair, and send it to the graphite
        server.

        :param metric: name of the metric
        :type prefix: string
        :param value: value of the metric
        :type prefix: float or int
        :param timestmap: epoch time of the event
        :type prefix: float or int
        :param formatter: option non-default formatter
        :type prefix: callable

        .. code-block:: python

          >>> g = init()
          >>> g.send("metric", 54)

        .. code-block:: python

          >>> g = init()
          >>> g.send(metric="metricname", value=73)

        """
        if formatter is None:
            formatter = self.formatter
        message = formatter(metric, value, timestamp)
        message = self. _presend(message)
        return self._dispatch_send(message)

    def send_dict(self, data, timestamp=None, formatter=None):
        """
        Format a dict of metric/values pairs, and send them all to the
        graphite server.

        :param data: key,value pair of metric name and metric value
        :type prefix: dict
        :param timestmap: epoch time of the event
        :type prefix: float or int
        :param formatter: option non-default formatter
        :type prefix: callable

        .. code-block:: python

          >>> g = init()
          >>> g.send_dict({'metric1': 54, 'metric2': 43, 'metricN': 999})

        """
        if formatter is None:
            formatter = self.formatter

        metric_list = []

        for metric, value in data.items():
            tmp_message = formatter(metric, value, timestamp)
            metric_list.append(tmp_message)

        message = "".join(metric_list)
        return self._dispatch_send(message)

    def send_list(self, data, timestamp=None, formatter=None):
        """

        Format a list of set's of (metric, value) pairs, and send them all
        to the graphite server.

        :param data: list of key,value pairs of metric name and metric value
        :type prefix: list
        :param timestmap: epoch time of the event
        :type prefix: float or int
        :param formatter: option non-default formatter
        :type prefix: callable

        .. code-block:: python

          >>> g = init()
          >>> g.send_list([('metric1', 54),('metric2', 43, 1384418995)])

        """
        if formatter is None:
            formatter = self.formatter

        if timestamp is None:
            timestamp = int(time.time())
        else:
            timestamp = int(timestamp)

        metric_list = []

        for metric_info in data:

            # Support [ (metric, value, timestamp), ... ] as well as
            # [ (metric, value), ... ].
            # If the metric_info provides a timestamp then use the timestamp.
            # If the metric_info fails to provide a timestamp, use the one
            # provided to send_list() or generated on the fly by time.time()
            if len(metric_info) == 3:
                (metric, value, metric_timestamp) = metric_info
            else:
                (metric, value) = metric_info
                metric_timestamp = timestamp

            tmp_message = formatter(metric, value, metric_timestamp)
            metric_list.append(tmp_message)

        message = "".join(metric_list)
        return self._dispatch_send(message)

    def enable_asynchronous(self):
        """Check if socket have been monkey patched by gevent"""

        def is_monkey_patched():
            try:
                from gevent import monkey, socket
            except ImportError:
                return False
            if hasattr(monkey, "saved"):
                return "socket" in monkey.saved
            return gevent.socket.socket == socket.socket

        if not is_monkey_patched():
            raise Exception("To activate asynchonoucity, please monkey patch"
                            " the socket module with gevent")
        return True


class GraphitePickleClient(GraphiteClient):

    def __init__(self, *args, **kwargs):
        # If the user has not given a graphite_port, then use the default pick
        # port.
        if 'graphite_port' not in kwargs:
            kwargs['graphite_port'] = default_graphite_pickle_port

        # TODO: Fix this hack and use super.
        # self = GraphiteClient(*args, **kwargs)  # noqa
        super(self.__class__, self).__init__(*args, **kwargs)

    def str2listtuple(self, string_message):
        "Covert a string that is ready to be sent to graphite into a tuple"

        if type(string_message).__name__ not in ('str', 'unicode'):
            raise TypeError("Must provide a string or unicode")

        if not string_message.endswith('\n'):
            string_message += "\n"

        tpl_list = []
        for line in string_message.split('\n'):
            line = line.strip()
            if not line:
                continue
            path, metric, timestamp = (None, None, None)
            try:
                (path, metric, timestamp) = line.split()
            except ValueError:
                raise ValueError(
                    "message must contain - metric_name, value and timestamp '%s'"
                    % line)
            try:
                timestamp = float(timestamp)
            except ValueError:
                raise ValueError("Timestamp must be float or int")

            tpl_list.append((path, (timestamp, metric)))

        if len(tpl_list) == 0:
            raise GraphiteSendException("No messages to send")

        payload = pickle.dumps(tpl_list)
        header = struct.pack("!L", len(payload))
        message = header + payload

        return message

    def _send(self, message):
        """ Given a message send it to the graphite server. """

        # An option to lowercase the entire message
        if self.lowercase_metric_names:
            message = message.lower()

        # convert the message into a pickled payload.
        message = self.str2listtuple(message)

        try:
            self.socket.sendall(message)
        # Capture missing socket.
        except socket.gaierror as error:
            raise GraphiteSendException(
                "Failed to send data to %s, with error: %s" %
                (self.addr, error))  # noqa

        # Capture socket closure before send.
        except socket.error as error:
            raise GraphiteSendException(
                "Socket closed before able to send data to %s, "
                "with error: %s" %
                (self.addr, error))  # noqa

        except Exception as error:
            raise GraphiteSendException(
                "Unknown error while trying to send data down socket to %s, "
                "error: %s" %
                (self.addr, error))  # noqa

        return "sent %d long pickled message" % len(message)


def init(init_type='plaintext_tcp', *args, **kwargs):
    """
    Create the module instance of the GraphiteClient.
    """
    global _module_instance
    reset()

    validate_init_types = ['plaintext_tcp', 'plaintext', 'pickle_tcp',
                           'pickle', 'plain']

    if init_type not in validate_init_types:
        raise GraphiteSendException(
            "Invalid init_type '%s', must be one of: %s" %
            (init_type, ", ".join(validate_init_types)))

    # Use TCP to send data to the plain text receiver on the graphite server.
    if init_type in ['plaintext_tcp', 'plaintext', 'plain']:
        _module_instance = GraphiteClient(*args, **kwargs)

    # Use TCP to send pickled data to the pickle receiver on the graphite
    # server.
    if init_type in ['pickle_tcp', 'pickle']:
        _module_instance = GraphitePickleClient(*args, **kwargs)

    return _module_instance


def send(*args, **kwargs):
    """ Make sure that we have an instance of the GraphiteClient.
    Then send the metrics to the graphite server.
    User consumable method.
    """
    if not _module_instance:
        raise GraphiteSendException(
            "Must call graphitesend.init() before sending")

    _module_instance.send(*args, **kwargs)
    return _module_instance


def send_dict(*args, **kwargs):
    """ Make sure that we have an instance of the GraphiteClient.
    Then send the metrics to the graphite server.
    User consumable method.
    """
    if not _module_instance:
        raise GraphiteSendException(
            "Must call graphitesend.init() before sending")
    _module_instance.send_dict(*args, **kwargs)
    return _module_instance


def send_list(*args, **kwargs):
    """ Make sure that we have an instance of the GraphiteClient.
    Then send the metrics to the graphite server.
    User consumable method.
    """
    if not _module_instance:
        raise GraphiteSendException(
            "Must call graphitesend.init() before sending")
    _module_instance.send_list(*args, **kwargs)
    return _module_instance


def reset():
    """ disconnect from the graphite server and destroy the module instance.
    """
    global _module_instance
    if not _module_instance:
        return False
    _module_instance.disconnect()
    _module_instance = None


def cli():
    """ Allow the module to be called from the cli. """
    import argparse

    parser = argparse.ArgumentParser(description='Send data to graphite')

    # Core of the application is to accept a metric and a value.
    parser.add_argument('metric', metavar='metric', type=str,
                        help='name.of.metric')
    parser.add_argument('value', metavar='value', type=int,
                        help='value of metric as int')

    args = parser.parse_args()
    metric = args.metric
    value = args.value

    graphitesend_instance = init()
    graphitesend_instance.send(metric, value)

if __name__ == '__main__':  # pragma: no cover
    cli()
